cmd/readiness/main.go

```diff
@@ -0,0 +1,75 @@
+/*
+Copyright 2025 The llm-d Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"flag"
+	"os"
+	"sync"
+
+	"github.com/llm-d-incubation/llm-d-fast-model-actuation/pkg/server/requester/probes"
+	"github.com/llm-d-incubation/llm-d-fast-model-actuation/pkg/server/requester/spi"
+	"sigs.k8s.io/controller-runtime/pkg/manager/signals"
+
+	"k8s.io/klog/v2"
+)
+
+func main() {
+	klog.InitFlags(nil)
+	flag.Parse()
+
+	// set up signals so we handle the shutdown signal gracefully
+	ctx := signals.SetupSignalHandler()
+	logger := klog.FromContext(ctx)
+
+	// Read ports from environment variables, fallback to defaults
+	probesPort := os.Getenv("PROBES_PORT")
+	if probesPort == "" {
+		probesPort = "8080"
+	}
+
+	spiPort := os.Getenv("SPI_PORT")
+	if spiPort == "" {
+		spiPort = "8081"
+	}
+
+	ipCh := make(chan string)
+
+	var wg sync.WaitGroup
+	wg.Add(2)
+
+	go func() {
+		defer wg.Done()
+
+		err := spi.Start(ctx, spiPort, ipCh)
+		if err != nil {
+			logger.Error(err, "failed to start requester SPI server")
+		}
+	}()
+
+	// Start the readiness probe server
+	go func() {
+		defer wg.Done()
+
+		err := probes.Start(ctx, probesPort, ipCh)
+		if err != nil {
+			logger.Error(err, "failed to start requester probes server")
+		}
+	}()
+
+	wg.Wait()
```

Comment:
This seems much more complicated than necessary to me. I see no need for two HTTP servers. I see no need for two additional packages. I see no need for the case of taking DOWN readiness. I see only a need for one file, one HTTP server, two entries in the mux (`/readyz` and `/dual-pod/v1/inference-server-ip`). Communication between the two handlers by an atomic that holds the IP (or not).

Reply:
2 servers for security reasons as clients are different. That way it's easy to setup network policies.

Inference server pods can restart which then can receive a new IP.

Reply:
A server-running Pod is not "normally" deleted --- except by the dual-pod controller when that deletion is desired. A Node can be deleted, which cascades to deletion of the Pods on it. IIRC, there are also unusual circumstances in which Kubernetes will evict a Pod from a Node.So we _do_ need to think about what happens when something besides the dual-pod controller deletes a server-running Pod, which I hadn't done.